// @okta/okta-angular/esm2015/src/okta/has-any-group.directive.js
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { OktaAuthStateService } from './services/auth-state.service';
export class OktaHasAnyGroupDirective {
    constructor(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    templateRef, viewContainer, authStateService) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.authStateService = authStateService;
    }
    set oktaHasAnyGroup(groups) {
        this.authStateService.hasAnyGroups(groups)
            .subscribe(isAuthorized => {
            // don't update UI if no state change
            if (isAuthorized === this.previousIsAuthorized) {
                return;
            }
            this.previousIsAuthorized = isAuthorized;
            this.viewContainer.clear();
            if (isAuthorized) {
                this.viewContainer.createEmbeddedView(this.templateRef);
            }
        });
    }
}
OktaHasAnyGroupDirective.decorators = [
    { type: Directive, args: [{ selector: '[oktaHasAnyGroup]' },] }
];
OktaHasAnyGroupDirective.ctorParameters = () => [
    { type: TemplateRef },
    { type: ViewContainerRef },
    { type: OktaAuthStateService }
];
OktaHasAnyGroupDirective.propDecorators = {
    oktaHasAnyGroup: [{ type: Input }]
};
