// @okta/okta-angular/esm5/src/okta/okta.guard.js
/*
 * Copyright (c) 2017-Present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */
import { __awaiter, __generator } from "tslib";
import { Injectable, Injector, Inject } from '@angular/core';
import { Router, NavigationStart } from '@angular/router';
import { filter } from 'rxjs/operators';
import { OktaAuth } from '@okta/okta-auth-js';
import { OKTA_CONFIG } from './models/okta.config';
var OktaAuthGuard = /** @class */ (function () {
    function OktaAuthGuard(config, oktaAuth, injector) {
        var _this = this;
        this.config = config;
        this.oktaAuth = oktaAuth;
        this.injector = injector;
        this.updateAuthStateListener = function (authState) {
            if (!authState.isAuthenticated) {
                _this.handleLogin(_this.state.url);
            }
        };
        this.onAuthRequired = this.config.onAuthRequired;
        // Unsubscribe updateAuthStateListener when route change
        var router = injector.get(Router);
        router.events.pipe(filter(function (e) { return e instanceof NavigationStart && _this.state && _this.state.url !== e.url; })).subscribe(function () {
            _this.oktaAuth.authStateManager.unsubscribe(_this.updateAuthStateListener);
        });
    }
    OktaAuthGuard.prototype.canLoad = function (route, segments) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var isAuthenticated, originalUri;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this.onAuthRequired = ((_a = route.data) === null || _a === void 0 ? void 0 : _a.onAuthRequired) || this.onAuthRequired;
                        return [4 /*yield*/, this.oktaAuth.isAuthenticated()];
                    case 1:
                        isAuthenticated = _b.sent();
                        if (isAuthenticated) {
                            return [2 /*return*/, true];
                        }
                        originalUri = segments[0].path;
                        return [4 /*yield*/, this.handleLogin(originalUri)];
                    case 2:
                        _b.sent();
                        return [2 /*return*/, false];
                }
            });
        });
    };
    /**
     * Gateway for protected route. Returns true if there is a valid accessToken,
     * otherwise it will cache the route and start the login flow.
     * @param route
     * @param state
     */
    OktaAuthGuard.prototype.canActivate = function (route, state) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var isAuthenticated;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        // Track states for current route
                        this.state = state;
                        this.onAuthRequired = ((_a = route.data) === null || _a === void 0 ? void 0 : _a.onAuthRequired) || this.onAuthRequired;
                        // Protect the route after accessing
                        this.oktaAuth.authStateManager.subscribe(this.updateAuthStateListener);
                        return [4 /*yield*/, this.oktaAuth.isAuthenticated()];
                    case 1:
                        isAuthenticated = _b.sent();
                        if (isAuthenticated) {
                            return [2 /*return*/, true];
                        }
                        return [4 /*yield*/, this.handleLogin(state.url)];
                    case 2:
                        _b.sent();
                        return [2 /*return*/, false];
                }
            });
        });
    };
    OktaAuthGuard.prototype.canActivateChild = function (route, state) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.canActivate(route, state)];
            });
        });
    };
    OktaAuthGuard.prototype.handleLogin = function (originalUri) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                // Store the current path
                this.oktaAuth.setOriginalUri(originalUri);
                if (this.onAuthRequired) {
                    this.onAuthRequired(this.oktaAuth, this.injector);
                }
                else {
                    this.oktaAuth.signInWithRedirect();
                }
                return [2 /*return*/];
            });
        });
    };
    OktaAuthGuard.decorators = [
        { type: Injectable }
    ];
    OktaAuthGuard.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [OKTA_CONFIG,] }] },
        { type: OktaAuth },
        { type: Injector }
    ]; };
    return OktaAuthGuard;
}());
export { OktaAuthGuard };
